cmd/snap-confine/snap-confine.apparmor.in

```diff
@@ -14,15 +14,15 @@
     # any abstractions
     /etc/ld.so.cache r,
     /etc/ld.so.preload r,
-    /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}ld-*.so mrix,
+    /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}{,atomics/}ld-*.so mrix,
```

Comment:
This rule means all of these could match for /lib:
* /lib/ld-...
* /lib/aarch64-linux-gnu/ld-...
* /lib/atomics/ld-...
* /lib/aarch64-linux-gnu/atomics/ld-...

Is that 3rd one valid? If not, these should be changed to: `/{,usr/}lib{,32,64,x32}/{,@{multiarch}/{,atomics/}}<rest of the rules>...`

Reply:
I _hope_ not but I really don't know the design of this feature or why a sub-directory was used (maybe it's an alternative system?)

I can restrict it to the suggestion you posted and we can open it up more if it comes back to bite us.

Reply:
The PR description mentions that it is only under the GNU triplet, so I'd feel more comfortable going with that until, like you said, something shows we need to do otherwise.

Reply:
> This rule means all of these could match for /lib:
> 
> * /lib/ld-...
> * /lib/aarch64-linux-gnu/ld-...
> * /lib/atomics/ld-...
> * /lib/aarch64-linux-gnu/atomics/ld-...
> 
> Is that 3rd one valid? If not, these should be changed to: `/{,usr/}lib{,32,64,x32}/{,@{multiarch}/{,atomics/}}<rest of the rule>...`

ld.so has a built-in search path. On multiarch enabled systems, it's not just the usual `/lib | /usr/lib` things, but also suffixed with multiarch triplet.

Then there are hwcaps with are appended to each of the search path. The longest one wins.

For example, on my machine:
```
$ LD_DEBUG=libs /lib64/ld-linux-x86-64.so.2 --inhibit-cache /usr/bin/true
     62126:	find library=libc.so.6 [0]; searching
     62126:	 search path=/lib/x86_64-linux-gnu/tls/haswell/x86_64:/lib/x86_64-linux-gnu/tls/haswell:/lib/x86_64-linux-gnu/tls/x86_64:/lib/x86_64-linux-gnu/tls:/lib/x86_64-linux-gnu/haswell/x86_64:/lib/x86_64-linux-gnu/haswell:/lib/x86_64-linux-gnu/x86_64:/lib/x86_64-linux-gnu:/usr/lib/x86_64-linux-gnu/tls/haswell/x86_64:/usr/lib/x86_64-linux-gnu/tls/haswell:/usr/lib/x86_64-linux-gnu/tls/x86_64:/usr/lib/x86_64-linux-gnu/tls:/usr/lib/x86_64-linux-gnu/haswell/x86_64:/usr/lib/x86_64-linux-gnu/haswell:/usr/lib/x86_64-linux-gnu/x86_64:/usr/lib/x86_64-linux-gnu:/lib/tls/haswell/x86_64:/lib/tls/haswell:/lib/tls/x86_64:/lib/tls:/lib/haswell/x86_64:/lib/haswell:/lib/x86_64:/lib:/usr/lib/tls/haswell/x86_64:/usr/lib/tls/haswell:/usr/lib/tls/x86_64:/usr/lib/tls:/usr/lib/haswell/x86_64:/usr/lib/haswell:/usr/lib/x86_64:/usr/lib		(system search path)
     62126:	  trying file=/lib/x86_64-linux-gnu/tls/haswell/x86_64/libc.so.6
     62126:	  trying file=/lib/x86_64-linux-gnu/tls/haswell/libc.so.6
     62126:	  trying file=/lib/x86_64-linux-gnu/tls/x86_64/libc.so.6
     62126:	  trying file=/lib/x86_64-linux-gnu/tls/libc.so.6
     62126:	  trying file=/lib/x86_64-linux-gnu/haswell/x86_64/libc.so.6
     62126:	  trying file=/lib/x86_64-linux-gnu/haswell/libc.so.6
     62126:	  trying file=/lib/x86_64-linux-gnu/x86_64/libc.so.6
     62126:	  trying file=/lib/x86_64-linux-gnu/libc.so.6
     62126:	
     62126:	
     62126:	calling init: /lib/x86_64-linux-gnu/libc.so.6
     62126:	
     62126:	
     62126:	initialize program: /usr/bin/true
     62126:	
     62126:	
     62126:	transferring control: /usr/bin/true
     62126:	
     62126:	
     62126:	calling fini: /usr/bin/true [0]
     62126:	
```

Note the full search path and how it has all the weird hwcaps subpaths. Just like
```
...
/lib/haswell/x86_64
/lib/haswell
/lib/x86_64
...
```
Are valid, I do expect `/lib/atomics` to be valid.

Obviously things shouldn't be silly enough to use all of these available paths. And we should be free to restrict things a little. At the moment, for atomics, I only anticipate to have `/usr/lib/aarch64-linux-gnu/atomics` to be used as a search path for extra libraries.

Reply:
Thank you for the fantastic explanation @xnox, I just learned bout hwcaps.